src/cata_ui.cpp

```diff
@@ -0,0 +1,1041 @@
+        x_offset += draw_subtab( win, get_ax() + x_offset, get_ay(), text[i], i == scroll ) + 2;
+    }
+}
+
+void ui_horizontal_list::set_text( std::vector<std::string> text )
+{
+    this->text = text;
+    size_t text_length = 1;
+
+    for( const auto &str : text ) {
+        text_length += utf8_width( str ) + 1;
+    }
+
+    set_rect( ui_rect( text_length, get_rect().size_y, get_rect().x, get_rect().y ) );
+}
+
+void ui_horizontal_list::scroll_left()
+{
+    scroll = ( scroll == 0 ? text.size() - 1 : scroll - 1 );
+}
+
+void ui_horizontal_list::scroll_right()
+{
+    scroll = ( scroll == text.size() - 1 ? 0 : scroll + 1 );
+}
+
+const std::string &ui_horizontal_list::current() const
+{
+    return text[scroll];
+}
+
+/////////////////////////////////////////////////////
+//THE FOLLOWING IS TEMPORARY/////////////////////////
+/////////////////////////////////////////////////////
+void label_test()
+{
+    auto lable1 = new ui_label("some", 0, 0, top_left);
+    lable1->text_color = c_red;
+    auto lable2 = new ui_label("anchored", 0, 0, center_center);
+    lable2->text_color = c_ltblue;
+    auto lable3 = new ui_label("labels", 0, 0, bottom_right);
+    lable3->text_color = c_ltgreen;
+
+    bordered_window win(31, 13, 50, 15);
+    win.add_child(lable1);
+    win.add_child(lable2);
+    win.add_child(lable3);
+    win.draw();
+}
+
+void tab_test()
+{
+    tabbed_window win(31, 14, 50, 15);
+    auto t_win1 = win.create_tab<ui_window>("tab 1");
+    auto label1 = new ui_label("window 1", 0, 0, center_center);
+    t_win1->add_child(label1);
+
+    auto t_win2 = win.create_tab<ui_window>("tab 2");
+    auto label2 = new ui_label("window 2", 0, 0, center_center);
+    t_win2->add_child(label2);
+
+    win.draw();
+
+    input_context ctxt;
+    ctxt.register_action("QUIT");
+    ctxt.register_action("PREV_TAB");
+    ctxt.register_action("NEXT_TAB");
+
+    while(true) {
+        const std::string action = ctxt.handle_input();
+
+        if(action == "PREV_TAB") {
+            win.previous_tab();
+        } else if(action == "NEXT_TAB") {
+            win.next_tab();
+        } else if(action == "QUIT") {
+            break;
+        }
+
+        win.draw();
+    }
+}
+
+void indicators_test()
+{
+    bordered_window win(31, 31, 50, 15);
+
+    auto hb = new health_bar(5, 0, 0, center_center);
+    win.add_child(hb);
+    hb->set_health_percentage(0.5);
+
+    auto si = new smiley_indicator(0, -1, center_center);
+    win.add_child(si);
+
+    win.draw();
+}
+
+void tile_panel_test()
+{
+    bordered_window win(31, 31, 50, 15);
+
+    auto tp = new tile_panel<ui_tile>(29, 29, 1, 1);
+    win.add_child(tp);
+
+    tp->set_tile(ui_tile('X', c_yellow), 5, 5);
+    tp->set_tile(ui_tile('X', c_yellow), 9, 5);
+    tp->set_tile(ui_tile('X', c_yellow), 7, 6);
+
+    win.draw();
+}
+
+void auto_border_test()
+{
+    auto_bordered_window win(51, 23, 50, 15);
+
+    win.add_child(new bordered_window(49, 10, 1, 1));
```

Comment:
Is there a way to add non bordered subwindows? Don't see many uses for fully bordered subwindows, since we don't have that much space to play with.

Maybe take a look at `draw_custom_border(...)` to be able to specify where the borders should be drawn.